drivers/gpio/gpio_mcp23s17.c

```diff
@@ -0,0 +1,520 @@
+/*
+ * Copyright (c) 2019 Geanix ApS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file Driver for MCP23S17 SPI-based GPIO driver.
+ */
+
+#include <errno.h>
+
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <misc/util.h>
+#include <gpio.h>
+#include <spi.h>
+
+#include "gpio_mcp23s17.h"
+
+#define LOG_LEVEL CONFIG_GPIO_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(gpio_mcp23s17);
+
+/**
+ * @brief Read both port 0 and port 1 registers of certain register function.
+ *
+ * Given the register in reg, read the pair of port 0 and port 1.
+ *
+ * @param dev Device struct of the MCP23S17.
+ * @param reg Register to read (the PORTA of the pair of registers).
+ * @param buf Buffer to read data into.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int read_port_regs(struct device *dev, u8_t reg,
+			   union mcp23s17_port_data *buf)
+{
+	struct mcp23s17_drv_data *const drv_data =
+	    (struct mcp23s17_drv_data * const)dev->driver_data;
+	int ret;
+
+	u8_t addr = MCP23S17_ADDR | MCP23S17_READBIT;
+	u8_t buffer_tx[4] = { addr, reg, 0, 0 };
+
+	const struct spi_buf tx_buf = {
+		.buf = buffer_tx,
+		.len = 4,
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+	struct spi_buf rx_buf[2];
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = 2,
+	};
+	rx_buf[0].buf = NULL;
+	rx_buf[0].len = 1;
+	rx_buf[1].buf = &buf->byte;
+	rx_buf[1].len = 1;
+
+	ret = spi_transceive(drv_data->spi, &drv_data->spi_cfg, &tx, &rx);
+	if (ret) {
+		LOG_DBG("spi_transceive FAIL %d\n", ret);
+		return ret;
+	}
+
+	LOG_DBG("MCP23S17: Read: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X",
+		reg, buf->byte[0], (reg + 1), buf->byte[1]);
+
+	return 0;
+}
+
+/**
+ * @brief Write both port 0 and port 1 registers of certain register function.
+ *
+ * Given the register in reg, write the pair of port 0 and port 1.
+ *
+ * @param dev Device struct of the MCP23S17.
+ * @param reg Register to write into (the PORTA of the pair of registers).
+ * @param buf Buffer to write data from.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int write_port_regs(struct device *dev, u8_t reg,
+			    union mcp23s17_port_data *buf)
+{
+	struct mcp23s17_drv_data *const drv_data =
+	    (struct mcp23s17_drv_data * const)dev->driver_data;
+	int ret;
+
+	u8_t addr = MCP23S17_ADDR;
+	u8_t buffer_tx[2] = { addr, reg };
+
+	const struct spi_buf tx_buf[2] = {
+		{
+			.buf = buffer_tx,
+			.len = 2,
+		},
+		{
+			.buf = &buf->byte,
```

Comment:
The data behind `buf->byte` are stored in machine order (little or big endian), having the union doesn't help here. To make sure the driver behavior is uniform among architectures we should use `sys_cpu_to_le16` function to translate data send to mcp23s17 (LSB always written to A register, MSB to B register). Code in `gpio_pca95xx.c` can be used as an example.

Similar comment holds true for `read_port_regs` function.

The `union mcp23s17_port_data` type can be removed from the driver code and replaced by `u16_t`.

Reply:
Agree, but it a larger change so I will see when I get time to fix